compiler/src/dotty/tools/dotc/core/TypeOps.scala

```diff
@@ -188,29 +192,82 @@ trait TypeOps { this: Context => // TODO: Make standalone object.
+      //
+      //   `A | C` under the constraints `B >: A` and `C <: B`
+      //
+      // where `A, B, C` are type parameters.
+      // Widening `A` to its upper bound would give `Any | C`, i.e. `Any`.
+      // But widening `C` first would give `A | B` and then `B`.
+      // So we need to widen `C` first. But how to decide this in general?
+      // In the algorithm below, we try to widen both sides (once), and then proceed as follows:
+      //
+      //  2.0. If no widening succeeds, proceed with step 3.
+      //  2.1. If only one widening succeeds, pick that one.
+      //  2.2. If the two widened types are in a subtype relationship, pick the smaller one.
+      //  2.3. If exactly one of the two types is a singleton type, pick that one.
+      //  2.4. If the widened tp1 is a supertype of tp2, pick widened tp1.
+      //  2.5. If the widened tp2 is a supertype of tp1, pick widened tp2.
+      //  2.6. Otherwise, pick tp1
+      //
+      // At steps 4-6 we lose possible solutions, since we have to make an
+      // arbitrary choice which side to widen. A better solution would look at
+      // the constituents of each operand (if the operand is an OrType again) and
+      // try to widen them selectively in turn. But this might lead to a combinatorial
+      // explosion of possibilities.
+      //
+      // Another approach could be to store information contained in lower bounds
+      // on both sides. So if `B >: A` we'd also record that `A <: B` and therefore
+      // widening `A` would yield `B` instead of `Any`, so we'd still be on the right track.
+      // This looks feasible if lower bounds are type parameters, but tricky if they
+      // are something else. We'd have to extract the strongest possible
+      // constraint over all type parameters that is implied by a lower bound.
+      // This looks related to an algorithmic problem arising in GADT matching.
+      //
+      // However, this alone is still not enough. There are other sources of incompleteness,
+      // for instance arising from mis-aligned refinements.
+      val tp1w = tp1 match {
+        case tp1: TypeProxy if !isClassRef(tp1) => tp1.superType.widenExpr
+        case _ => tp1
+      }
+      val tp2w = tp2 match {
+        case tp2: TypeProxy if !isClassRef(tp2) => tp2.superType.widenExpr
+        case _ => tp2
+      }
+      if ((tp1w ne tp1) || (tp2w ne tp2)) {
+        val isSingle1 = tp1.isInstanceOf[SingletonType]
+        val isSingle2 = tp2.isInstanceOf[SingletonType]
+        return {
+          if (tp2w eq tp2) orDominator(tp1w | tp2)
+          else if (tp1w eq tp1) orDominator(tp1 | tp2w)
+          else if (tp1w frozen_<:< tp2w) orDominator(tp1w | tp2)
+          else if (tp2w frozen_<:< tp1w) orDominator(tp1 | tp2w)
+          else if (isSingle1 && !isSingle2) orDominator(tp1w | tp2)
+          else if (isSingle2 && !isSingle1) orDominator(tp1 | tp2w)
+          else if (tp1 frozen_<:< tp2w) tp2w
```

Comment:
Is a case for 2.5 looking like `else if (tp2 frozen_<:< tp1w) tp1w` missing here ?